bee-message/tests/tagged_data_payload.rs

```diff
@@ -0,0 +1,128 @@
+// Copyright 2020-2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use bee_message::{payload::tagged_data::TaggedDataPayload, Error, Message};
+use bee_test::rand::bytes::rand_bytes;
+
+use packable::{
+    bounded::{TryIntoBoundedU32Error, TryIntoBoundedU8Error},
+    error::UnpackError,
+    PackableExt,
+};
+
+#[test]
+fn kind() {
+    assert_eq!(TaggedDataPayload::KIND, 5);
+}
+
+#[test]
+fn new_valid() {
+    let tag = rand_bytes(64);
+    let data = vec![0x42, 0xff, 0x84, 0xa2, 0x42, 0xff, 0x84, 0xa2];
+    let tagged_data = TaggedDataPayload::new(tag.clone(), data.clone()).unwrap();
+
+    assert_eq!(tagged_data.tag(), &tag);
+    assert_eq!(tagged_data.data(), &data);
+}
+
+#[test]
+fn new_valid_empty_data() {
+    let tag = rand_bytes(64);
+    let data = vec![];
+    let tagged_data = TaggedDataPayload::new(tag.clone(), data.clone()).unwrap();
+
+    assert_eq!(tagged_data.tag(), &tag);
+    assert_eq!(tagged_data.data(), &data);
+}
+
+#[test]
+fn new_valid_padded() {
+    let tag = rand_bytes(32);
+    let data = vec![];
+    let tagged_data = TaggedDataPayload::new(tag.clone(), data.clone()).unwrap();
+
+    assert_eq!(tagged_data.tag(), &tag);
+    assert_eq!(tagged_data.data(), &data);
+}
+
+#[test]
+fn new_invalid_tag_length_less_than_min() {
+    assert!(matches!(
+        TaggedDataPayload::new(vec![], vec![0x42, 0xff, 0x84, 0xa2, 0x42, 0xff, 0x84, 0xa2]),
+        Err(Error::InvalidTagLength(TryIntoBoundedU8Error::Invalid(0)))
+    ));
+}
+
+#[test]
+fn new_invalid_tag_length_more_than_max() {
+    assert!(matches!(
+        TaggedDataPayload::new(rand_bytes(65), vec![0x42, 0xff, 0x84, 0xa2, 0x42, 0xff, 0x84, 0xa2]),
+        Err(Error::InvalidTagLength(TryIntoBoundedU8Error::Invalid(65)))
+    ));
+}
+
+#[test]
+fn new_invalid_data_length_more_than_max() {
+    assert!(matches!(
+        TaggedDataPayload::new(rand_bytes(32), vec![0u8; Message::LENGTH_MAX + 42]),
+        Err(Error::InvalidTaggedDataLength(TryIntoBoundedU32Error::Invalid(l))) if l == Message::LENGTH_MAX as u32 + 42
+    ));
+}
+
+#[test]
+fn packed_len() {
+    let tagged_data =
+        TaggedDataPayload::new(rand_bytes(10), vec![0x42, 0xff, 0x84, 0xa2, 0x42, 0xff, 0x84, 0xa2]).unwrap();
+
+    assert_eq!(tagged_data.packed_len(), 10 + 1 + 4 + 8);
```

Comment:
Just to be sure: `1` is the prefix for the tag length and the `4` is the prefix for the data length?

Reply:
Yeah, it's confusing that 1 and 10 are reversed, I'll change